File: sol/solution/0101-0200/0110/Solution.erl
%% Definition for a binary tree node.
%%
%% -record(tree_node, {val = 0 :: integer(),
%%                     left = null  :: 'null' | #tree_node{},
%%                     right = null :: 'null' | #tree_node{}}).
-spec is_balanced(Root :: #tree_node{} | null) -> boolean().
is_balanced(null) -> true;
is_balanced(Root) -> check(Root) =/= -1.
check(null) -> 0;
check(#tree_node{left = Left, right = Right}) ->
    L = check(Left),
    case L of
        -1 -> -1;
        _ ->
            R = check(Right),
            case R of
                -1 -> -1;
                _ -> if abs(L - R) > 1 -> -1; true -> max(L, R) + 1 end
            end
    end.

File: sol/solution/0101-0200/0110/Solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
function isBalanced(root: TreeNode | null): boolean {
    function check(node: TreeNode | null): number {
        if (!node) return 0;
        const left = check(node.left);
        if (left === -1) return -1;
        const right = check(node.right);
        if (right === -1) return -1;
        if (Math.abs(left - right) > 1) return -1;
        return Math.max(left, right) + 1;
    }
    return check(root) !== -1;
}
